### PR TITLE
Wayland/virtual keyboard

### PR DESCRIPTION
**Pull request by rano-oss**
This adds support for virtual keyboard, ~~it needs a new version of wayland-protocols-misc for it to be able to be merged. The pr probably needs some reviewing first anyway. ;)~~
edit: Everything is tested and implemented, should be ready :')

### CODE REVIEW

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (i509VCB) [inline src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: I guess the protocol does not mark the format u32 as an enum...

Comment (rano-oss) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: Unfortunately not

Comment (i509VCB) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: We should probably open an issue for that then where ever this protocol lives

Comment (HeavyRain266) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: @i509VCB This is part of wayland-protocols

Comment (i509VCB) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: No? This was posted on the mailing list but never integrated into wayland-protocols. I think purism is the source of the protocol.

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (i509VCB) [inline src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:60]: - A ZwpVirtualKeyboardV1 has a strong reference to the VirtualKeyboardUserData.
- A VirtualKeyboardUserData contains a `VirtualKeyboard`
- A VirtualKeyboard contains a `Vec<ZwpVirtualKeyboardV1>`

I think it's pretty likely we have a reference cycle which will leak memory.

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (i509VCB) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: I guess we should at least note why this is weird.

Something like:
```
This should be wl_keyboard::KeymapFormat::XkbV1, but the protocol does not state the parameter is an enum.
```

Comment (PolyMeilex) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: In addition to mailing list it also has MR open now: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/11

Comment (i509VCB) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: The MR indeed has the enum label

Comment (rano-oss) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:77]: The alternative would be to use the new version, basically being a forerunner?

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (rano-oss) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:72]: VirtualKeyboard now contains just a u8, I wasn't using the reference for anything anyway, just needed to count virtual keyboard instances.

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (rano-oss) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:91]: added comment :)

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (PolyMeilex) [inline src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:119]: Not saying if this is right or not, but what does happen if someone presses a key on a physical keyboard while the virtual keyboard is open? Won't the map be wrong?

Comment (rano-oss) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs:119]: yes, it will be the keymap from the virtual keyboard. So yes the map will be "wrong". The alternative would be to keep all keymaps and set the keymaps on all clients before sending the key.... That would be a lot of work.
In sway they only change it when it is different. My take on this, is that in practice virtual keyboards will set the same keymap as the one set on the compositor(aka no change), also input methods will probably set a different one in many cases but that will be intended behavior and thus the physical keyboard will output the "correct" map.
I do set the keymap back to the original one when the virtual one is dropped, which I would guess at least makes it somewhat predictable...
I am open for suggestions :smile:

File: src/wayland/seat/keyboard.rs
Comment (rano-oss) [reply src/wayland/seat/keyboard.rs]: yes, you are correct :sweat_smile:

**Comment by rano-oss** [comment]
Seems that github actions is not allowing clippy to run properly here

**Comment by PolyMeilex** [comment]
> Seems that github actions is not allowing clippy to run properly here?

Smithay build broke after new wayland-rs 0.30 beta got released, proper fix is #752

**Comment by rano-oss** [comment]
I guess I will wait for that then :smiling_face_with_tear:

File: src/input/keyboard/mod.rs
Comment (Drakulix) [inline src/input/keyboard/mod.rs:454]: I think this should rather be a warning

**Review by Drakulix — COMMENTED** [review]
Looks mostly fine from a code point of view, but I am not sure on the behavior from the protocol side.

Seems like keymap handling is real hassle and I don't know, if this is handled in the best way. I would like input from @i509VCB or @PolyMeilex before merging this.

File: src/wayland/virtual_keyboard/virtual_keyboard_handle.rs
Comment (PolyMeilex) [reply src/wayland/virtual_keyboard/virtual_keyboard_handle.rs]: Sounds like a reasonable compromise, I'm fine with that.

**Review by PolyMeilex — APPROVED** [review]
My concerns about keymap handling were addressed so it's good to go as far as I'm concerned.

File: src/input/keyboard/mod.rs
Comment (rano-oss) [reply src/input/keyboard/mod.rs:454]: Should I change this or merge and create a new shorter pr? :sweat_smile:

Comment (rano-oss) [reply src/input/keyboard/mod.rs]: Figured I might as well :smile_cat:

**Comment by Drakulix** [comment]
Sadly you broke CI now. 😅
I'll merge once its green again.

**Comment by rano-oss** [comment]
Just a minor hiccup :see_no_evil:

**Pull request merged**
